### ISSUE TITLE
Clarification on struct-of-arrays vs arrays-of-structs

### ISSUE THREAD

**Issue by maxnoe** [scikit-hep/awkward#649]
In the [C++ doxygen API](https://awkward-array.readthedocs.io/en/latest/_static/index.html), the `RecordArray` is explained as 

> ak::RecordArray: represents an array of records in "struct of arrays" form.

Shouldn't it be 

> "array of structs" 

?

**Comment by jpivarski** [comment]
This is correct. It means that although the data are logically presented as an array of records:

```
[{"x": 1, "y": 1.1}, {"x": 2, "y": 2.2}, {"x": 3, "y": 3.3}]
```

they're internally stored as a struct of arrays:

```
x: [1, 2, 3]
y: [1.1, 2.2, 3.3]
```

This is a general principle for all data structures in Awkward Array (as well as Apache Arrow and Parquet, other "columnar" data representations). It has a name, "AoS vs SoA," in the case of records, but this technique can be generalized to variable-length lists, missing data, heterogenous data, etc.

To see the columnar form of any Awkward Array in Python, look at its `layout`:

```python
>>> array.layout
```

A representation of its in-memory layout is presented in an XML format (intended for readability).

**Comment by jpivarski** [comment]
I really want to convert this into a Discussion, but it looks like GitHub removed that button! Oh well.

**Comment by maxnoe** [comment]
Thanks for the clarification. At least to me it was unclear that the first part in the sentence refers to the semantics and the second to the in-memory data structure.

**Comment by jpivarski** [comment]
Oh, that helps. I'll do this:

```diff
--- a/docs-sphinx/index.rst
+++ b/docs-sphinx/index.rst
@@ -157,7 +157,7 @@ Navigation
    * :doc:`ak.layout.RegularArray`: splits its nested content into equal-length lists.
    * :doc:`ak.layout.ListArray`: splits its nested content into variable-length lists with full generality (may use its content non-contiguously, overlapping, or out-of-order).
    * :doc:`ak.layout.ListOffsetArray`: splits its nested content into variable-length lists, assuming contiguous, non-overlapping, in-order content.
-   * :doc:`ak.layout.RecordArray`: represents an array of records in "struct of arrays" form.
+   * :doc:`ak.layout.RecordArray`: represents a logical array of records with a "struct of arrays" layout in memory.
    * :doc:`ak.layout.Record`: represents a single record (not a subclass of :doc:`ak.layout.Content` in Python).
    * :doc:`ak.layout.IndexedArray`: rearranges and/or duplicates its content by lazily applying an integer index.
    * :doc:`ak.layout.IndexedOptionArray`: same as :doc:`ak.layout.IndexedArray` with missing values as negative indexes.
```

I only found this text on the landing page for the Sphinx Python API, though, not the Doxygen C++ API.

**Comment by jpivarski** [comment]
PR #650. If you have another suggestion for changing the wording (before the tests pass), I'll edit that PR.

Thanks!

**Comment by maxnoe** [comment]
Thanks, that is much clearer!

I saw it it in the bullet point list here in the doxygen c++ docs: https://awkward-array.readthedocs.io/en/latest/_static/index.html

**Comment by maxnoe** [comment]
Generated from this file here I think: https://github.com/scikit-hep/awkward-1.0/blob/main/docs-doxygen/index.md,

https://github.com/scikit-hep/awkward-1.0/pull/651

**Comment by jpivarski** [comment]
Thanks: my `grep` wasn't wide-ranging enough. I've fixed this instance, too (ddbcd74d66d66084f399b6983137a6aa05b4b844).
